kalite/topic_tools/content_recommendation.py

```diff
@@ -0,0 +1,613 @@
+   
+    ''' Need to add some more analysis to ensure that the user has not done these exercises yet'''
+
+    return get_exercises_from_topics(data)
+
+
+
+
+
+
+
+
+
+##################################### GENERAL HELPER FUNCTIONS ############################################
+
+''' TODO '''
+# Returns the most recent subtopic id that the given user has started and/or completed.
+def get_most_recent_subtopic(user):
+    return 'early-math' #dummy return
+
+
+#Given a list of subtopic/topic ids, returns an ordered list of the first 5 exercise ids under those ids
+def get_exercises_from_topics(topicId_list):
+    exs = []
+    for topic in topicId_list:
+        exercises = get_topic_exercises(topic)[:5] #can change this line to allow for more to be recommended
+        for e in exercises:
+            exs += [e['id']] #only add the id to the list
+
+    return exs
+
+
+#given a facility user model, return the most recent exercise ids that are still in-progress
+def get_most_recent_incomplete_exercises(user):
+    exercises_by_user = ExerciseLog.objects.filter(user=user)
+    exercises_by_user = ExerciseLog.objects.filter(id__lte=1) #temp
+    
+    #sorted by completion time in descending order (most recent first)
+    sorted_exercises = sorted(exercises_by_user, key=lambda student: student.completion_timestamp, reverse=True)
+
+    exercise_list = []
+
+    for exercise in sorted_exercises:
+        if exercise.complete == False:                  #only look for incomplete
+            exercise_list.append(exercise.exercise_id)  #append to list
+
+    return exercise_list                                #most recent + incomplete
+
+
+
+#given a facility user model, return the most recent exercise ids - incomplete AND complete
+def get_most_recent_exercises(user):
+    exercises_by_user = ExerciseLog.objects.filter(user=user)
+    
+    #sorted by completion time in descending order (most recent first)
+    sorted_exercises = sorted(exercises_by_user, key=lambda student: student.completion_timestamp, reverse=True)
+
+    return sorted_exercises   
+
+
+
+#given a user type (can be null), group id, and a facility name, return all users in that group
+#calls the already defined function in facility module
+def get_users_in_group(user_type, group_id, facility):
+    return kalite.facility.get_users_from_group(user_type, group_id, facility)
+
+
+###################################### BEGIN NEAREST NEIGHBORS ############################################
+
+### MULTI-PURPOSE NEAREST NEIGHBORS ALGORITH, USE AS YOU PLEASE ###
+### THE MAIN THING TO REMEMBER IS THAT get_recommended_exercises(subtopic) IS THE MAIN FUNCTION TO CALL ###
+
+###
+# Returns a dictionary with each subtopic and their related
+# topics.
+#
+###
+def generate_recommendation_data():
+
+    #hardcoded data, each subtopic is the key with its related subtopics and current courses as the values. Not currently in use.
+    data_hardcoded = {
```

Comment:
We should probably put this into the data/khan folder as a json blob that we can load in for this.